File: src/easy/t1/solve.go
package main

import "fmt"

/*
* @Author: zouyx
* @Email:
* @Date:   2024/11/8 下午5:54
* @Package:
 */

func main() {
	num1 := []int{4, 5, 6, 0, 0, 0}
	m := 3
	num2 := []int{1, 2, 3}
	n := 3
	merge(num1, m, num2, n)
	fmt.Println(num1)
}

// 这一题是典型的双指针
func merge(nums1 []int, m int, nums2 []int, n int) {
	var (
		pm = m - 1
		pn = n - 1
	)
	// 最坏的情况下时间复杂度 O(M+N)  最好的情况下 O(N)
	for i := m + n - 1; pm >= 0 && pn >= 0; i-- {
		if nums1[pm] > nums2[pn] {
			nums1[i] = nums1[pm]
			pm--
		} else {
			nums1[i] = nums2[pn]
			pn--
		}
	}
	fmt.Println(pn)

	//  最坏的情况下 时间复杂度 O(M)  最好的情况下 O(o)
	for i := 0; i <= pn; i++ {
		nums1[i] = nums2[i]
	}
}

File: src/easy/t1/solve.rs
/*
* @Author: zouyx
* @Email: zouyx@knowsec.com
* @Date:   2024/11/8 下午6:05
* @Package:
 */
pub fn merge(nums1: &mut [i32], m: i32, nums2: &mut [i32], n: i32) {
    let mut pm = m - 1;
    let mut pn = n - 1;
    let mut p = m + n - 1;
    while pm >= 0 && pn >= 0 {
        if nums1[pm as usize] > nums2[pn as usize] {
            nums1[p as usize] = nums1[pm as usize];
            pm -= 1
        } else {
            nums1[p as usize] = nums2[pn as usize];
            pn -= 1
        }
        p -= 1
    }
    for i in 0..pn + 1 {
        nums1[i as usize] = nums2[i as usize]
    }
}

#[cfg(test)]
mod test {
    use super::merge;
    #[test]
    fn easy_t1_solve() {
        let mut nums1 = vec![1, 2, 3, 0, 0, 0];
        let m = 3;
        let mut nums2 = vec![4, 5, 6];
        let n = 3;
        merge(&mut nums1, m, &mut nums2, n);
        assert_eq!(nums1, vec![1, 2, 3, 4, 5, 6])
    }
}
